src/Solve.cpp

```diff
@@ -0,0 +1,627 @@
+                        } else if (is_negative_const(mul_a->b)) {
+                            expr = mutate(mul_a->a >= div);
+                        } else {
+                            fail(Cmp::make(a, b));
+                        }
+                    } else if (is_lt) {
+                        if (is_positive_const(mul_a->b)) {
+                            expr = mutate(mul_a->a < (b + (mul_a->b - 1)) / mul_a->b);
+                        } else if (is_negative_const(mul_a->b)) {
+                            expr = mutate(mul_a->a > (b - (mul_a->b + 1)) / mul_a->b);
+                        } else {
+                            fail(Cmp::make(a, b));
+                        }
+                    } else if (is_gt) {
+                        if (is_positive_const(mul_a->b)) {
+                            expr = mutate(mul_a->a > div);
+                        } else if (is_negative_const(mul_a->b)) {
+                            expr = mutate(mul_a->a < div);
+                        } else {
+                            fail(Cmp::make(a, b));
+                        }
+                    } else if (is_ge) {
+                        if (is_positive_const(mul_a->b)) {
+                            expr = mutate(mul_a->a >= (b + (mul_a->b - 1)) / mul_a->b);
+                        } else if (is_negative_const(mul_a->b)) {
+                            expr = mutate(mul_a->a <= (b - (mul_a->b + 1)) / mul_a->b);
+                        } else {
+                            fail(Cmp::make(a, b));
+                        }
+                    }
+                }
+            }
+        } else if (a_uses_var && b_uses_var) {
+            // Convert to f(x) - g(x) == 0 and let the subtract mutator clean up.
+            expr = mutate(Cmp::make(a - b, 0));
+        }
+
+        if (!expr.defined()) {
+            if (a.same_as(op->a) && b.same_as(op->b)) {
+                expr = op;
+            } else {
+                expr = Cmp::make(a, b);
+            }
+        }
+    }
+
+    void visit(const LT *op) {
+        visit_cmp<LT, GT>(op);
+    }
+
+    void visit(const LE *op) {
+        visit_cmp<LE, GE>(op);
+    }
+
+    void visit(const GE *op) {
+        visit_cmp<GE, LE>(op);
+    }
+
+    void visit(const GT *op) {
+        visit_cmp<GT, LT>(op);
+    }
+
+    void visit(const EQ *op) {
+        visit_cmp<EQ, EQ>(op);
+    }
+
+    void visit(const NE *op) {
+        visit_cmp<NE, NE>(op);
+    }
+
+    void visit(const Variable *op) {
+        if (op->name == var) {
+            uses_var = true;
+            expr = op;
+        } else if (scope.contains(op->name)) {
+            CacheEntry e = scope.get(op->name);
+            expr = e.expr;
+            uses_var = uses_var || e.uses_var;
+        } else if (external_scope.contains(op->name)) {
+            Expr e = external_scope.get(op->name);
+            // Expressions in the external scope haven't been solved
+            // yet.
+            expr = mutate(e);
```

Comment:
Calling mutate on the expression from the external scope will put a solved version of it in the cache on first use, and pull it out of the cache on subsequent uses.